src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -29,8 +30,8 @@ internal class AggregatorStore
         private readonly object lockKeyValue2MetricAggs = new object();
 
         // Two-Level lookup. TagKeys x [ TagValues x Metrics ]
-        private readonly Dictionary<string[], Dictionary<object[], int>> keyValue2MetricAggs =
-            new Dictionary<string[], Dictionary<object[], int>>(new StringArrayEqualityComparer());
+        private readonly ConcurrentDictionary<string[], ConcurrentDictionary<object[], int>> keyValue2MetricAggs =
```

Comment:
@cijothomas I think if we switch the inner dictionary to `Hashtable` we can eliminate a couple of the challenges below while preserving the perf. `Hashtable` has the unique [thread-safety gurantee](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-5.0#thread-safety) that it is safe for many readers and only needs a lock for writing. Similar to `ConcurrentDictionary` but you can self-manage the lock. Why this is good is below between TryGet & TryAdd we do a bunch of logic where multiple threads can race. `Hashtable` takes care of that issue because we'll manage the lock. I think it simplifies the code and we don't have to worry about gaps in the array, or a losing thread having to spin wait. Something like this...

```csharp
        // Two-Level lookup. TagKeys x [ TagValues x Metrics ]
        private readonly ConcurrentDictionary<string[], Hashtable> keyValue2MetricAggs =
            new ConcurrentDictionary<string[], Hashtable>(new StringArrayEqualityComparer());

        internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> tags)
        {
            int len = tags.Length;
            if (len == 0)
            {
                if (this.metrics[0].StartTime == default)
                {
                    var dt = DateTimeOffset.UtcNow;
                    this.metrics[0] = new MetricPoint(this.aggType, dt, null, null);
                }

                return 0;
            }

            var storage = ThreadStaticStorage.GetStorage();

            storage.SplitToKeysAndValues(tags, out var tagKey, out var tagValue);

            if (len > 1)
            {
                Array.Sort<string, object>(tagKey, tagValue);
            }

            string[] seqKey = null;

            // GetOrAdd by TagKey at 1st Level of 2-level dictionary structure.
            // Get back a Dictionary of [ Values x Metrics[] ].
            if (!this.keyValue2MetricAggs.TryGetValue(tagKey, out var value2metrics))
            {
                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
                seqKey = new string[len];
                tagKey.CopyTo(seqKey, 0);

                value2metrics = new Hashtable(new ObjectArrayEqualityComparer()); // Need to implement non-generic IEqualityComparer for this to work
                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
                {
                    // Some other thread added value2metrics, take it.
                    // The one created by this thread is wasted.
                    this.keyValue2MetricAggs.TryGetValue(tagKey, out value2metrics);
                }
            }

            // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
            // Get back Metrics[].
            if (!(value2metrics[tagValue] is int aggregatorIndex))
            {
                lock (value2metrics)
                {
                    aggregatorIndex = (value2metrics[tagValue] as int?) ?? -1;
                    if (aggregatorIndex == -1)
                    {
                        this.metricPointIndex++;
                        aggregatorIndex = this.metricPointIndex;
                        if (aggregatorIndex >= MaxMetricPoints)
                        {
                            // sorry! out of data points.
                             return -1;
                        }

                        // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.

                        if (seqKey == null)
                        {
                            seqKey = new string[len];
                            tagKey.CopyTo(seqKey, 0);
                        }

                        var seqVal = new object[len];
                        tagValue.CopyTo(seqVal, 0);

                        value2metrics[tagValue] = aggregatorIndex;

                        ref var metricPoint = ref this.metrics[aggregatorIndex];
                        if (metricPoint.StartTime == default)
                        {
                            var dt = DateTimeOffset.UtcNow;
                            metricPoint = new MetricPoint(this.aggType, dt, seqKey, seqVal);
                        }
                    }
                }
            }

            return aggregatorIndex;
        }
```

A benchmark showing Dictionary vs ConcurrentDictionary vs Hashtable...

|               Method |      Mean |     Error |    StdDev |    Median |  Gen 0 | Allocated |
|--------------------- |----------:|----------:|----------:|----------:|-------:|----------:|
|           Dictionary | 445.03 us | 19.546 us | 57.632 us | 424.03 us | 0.9766 |     11 KB |
| ConcurrentDictionary |  56.91 us |  1.115 us |  1.735 us |  56.19 us | 0.5493 |      5 KB |
|            Hashtable |  56.35 us |  0.713 us |  0.667 us |  56.01 us | 0.5493 |      5 KB |

<details>
<summary>Benchmark code</summary>

```csharp
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using BenchmarkDotNet.Attributes;

namespace Benchmarks
{
    [MemoryDiagnoser]
    public class DictionaryBenchmarks
    {
        private readonly string keys = string.Empty;
        private readonly object values = new object();

        private readonly Dictionary<string, Dictionary<object, int>> dictionary = new Dictionary<string, Dictionary<object, int>>();
        private readonly ConcurrentDictionary<string, ConcurrentDictionary<object, int>> concurrentDictionary = new ConcurrentDictionary<string, ConcurrentDictionary<object, int>>();
        private readonly Hashtable hashtable = new Hashtable();

        [Benchmark]
        public int Dictionary()
        {
            int count = 0;

            Parallel.For(0, 1000, (i) =>
            {
                Dictionary<object, int> inner;
                lock (this.dictionary)
                {
                    if (!this.dictionary.TryGetValue(this.keys, out inner))
                    {
                        inner = new Dictionary<object, int>();
                        this.dictionary[this.keys] = inner;
                    }
                }

                int value;
                lock (inner)
                {
                    if (!inner.TryGetValue(this.values, out value))
                    {
                        value = 0;
                        inner[this.values] = value;
                    }
                }

                Interlocked.Add(ref count, value);
            });

            return count;
        }

        [Benchmark]
        public int ConcurrentDictionary()
        {
            int count = 0;

            Parallel.For(0, 1000, (i) =>
            {
                if (!this.concurrentDictionary.TryGetValue(this.keys, out ConcurrentDictionary<object, int> inner))
                {
                    inner = new ConcurrentDictionary<object, int>();
                    this.concurrentDictionary[this.keys] = inner;
                }

                if (!inner.TryGetValue(this.values, out var value))
                {
                    value = 0;
                    inner[this.values] = value;
                }

                Interlocked.Add(ref count, value);
            });

            return count;
        }

        [Benchmark]
        public int Hashtable()
        {
            int count = 0;

            Parallel.For(0, 1000, (i) =>
            {
                if (!(this.hashtable[this.keys] is Hashtable inner))
                {
                    lock (this.hashtable)
                    {
                        inner = this.hashtable[this.keys] as Hashtable;
                        if (inner == null)
                        {
                            inner = new Hashtable();
                            this.hashtable[this.keys] = inner;
                        }
                    }
                }

                if (!(inner[this.values] is int value))
                {
                    lock (inner)
                    {
                        if (!(inner[this.values] is int innerValue))
                        {
                            value = 0;
                            inner[this.values] = value;
                        }
                        else
                        {
                            value = innerValue;
                        }
                    }
                }

                Interlocked.Add(ref count, value);
            });

            return count;
        }
    }
}
```
</details>

Reply:
The nice thing about Hashtable/Hashset is that concurrent reads don't need lock.
I'm not sure if Hashtable has the same perf problem as Hashset, might need to confirm the .NET Framework perf.

https://github.com/open-telemetry/opentelemetry-dotnet/issues/708#issuecomment-636540689

Reply:
FWIW, here are those benchmarks from above run on net462, netcoreapp3.1, & net5.0:

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1237 (21H1/May2021Update)
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4400.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4400.0), X64 RyuJIT


|               Method |      Mean |    Error |    StdDev |  Gen 0 | Allocated |
|--------------------- |----------:|---------:|----------:|-------:|----------:|
|           Dictionary | 418.63 us | 8.223 us | 17.876 us |      - |      3 KB |
| ConcurrentDictionary |  70.61 us | 1.275 us |  1.193 us | 0.3662 |      3 KB |
|            Hashtable |  74.55 us | 1.427 us |  1.698 us | 0.3662 |      3 KB |

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1237 (21H1/May2021Update)
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.100-rc.1.21458.32
  [Host]     : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT
  DefaultJob : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT


|               Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|--------------------- |----------:|----------:|----------:|-------:|----------:|
|           Dictionary | 498.57 us | 23.672 us | 69.796 us | 0.9766 |     10 KB |
| ConcurrentDictionary |  65.73 us |  1.155 us |  1.375 us | 0.4883 |      5 KB |
|            Hashtable |  63.17 us |  1.242 us |  1.525 us | 0.4883 |      5 KB |

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1237 (21H1/May2021Update)
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.100-rc.1.21458.32
  [Host]     : .NET 5.0.10 (5.0.1021.41214), X64 RyuJIT
  DefaultJob : .NET 5.0.10 (5.0.1021.41214), X64 RyuJIT


|               Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|--------------------- |----------:|----------:|----------:|-------:|----------:|
|           Dictionary | 485.20 us | 20.901 us | 61.627 us | 0.9766 |     10 KB |
| ConcurrentDictionary |  61.81 us |  1.224 us |  1.457 us | 0.4883 |      5 KB |
|            Hashtable |  62.29 us |  1.236 us |  1.323 us | 0.4883 |      5 KB |

Very interesting that on .NET Core the allocations have gone way up!